determined-ai/determined
# feat: Add Antd utilities to UI Kit [WEB-1726]

## Description

This PR adds several utilities to the UI Kit and provides support for dynamic theming within the application. The functionality has been accomplished by implementing the following:

1. `UIProvider` - Refactored to take `theme` and `darkMode` props. All children of this will inherit the theme from this component allowing for dynamic theming.
2. `ThemeProvider` - Provider for stateUI context 
3. `useTheme`: A new hook that encompasses all of our internal business logic for determining the correct theme and mode for the application. 


Some component theming, such as primary Button colors are still hardcoded using an `Antd` design token as can be seen (https://github.com/determined-ai/determined/pull/8219/files#diff-754214362accf7d46eebd6f1354cdad9c155b15b1fed8aeef66b1c92ac764d8bR199)[here]. If we need to change this behavior then that can be discussed and implemented. 


Usage: 
UI Kit users will wrap their application with a `ThemeProvider`. Afterwards, the theme and mode can be set using a `UIProvider`, via the `theme` and `mode` exposed by `useUI` or users can implement custom logic on a per page basis. In our application's case we use a `useTheme` hook to correctly set the `theme` and `mode`.

Notes  
1. `DarkLight` usage has been removed from usage throughout the application and replaced by a `isDarkMode` boolean. 2. `useInitApi` was removed because theming is now dependent on components inheriting styling from a parent `UIProvider`. The `useInitApi` logic caused the new `notifications.config` to not be applied to notifications.  
3. Some basic html elements and free form text in the application are styled solely based on the `color-scheme` on the document root which is the reasoning [for this](https://github.com/determined-ai/determined/pull/8219/files#diff-754214362accf7d46eebd6f1354cdad9c155b15b1fed8aeef66b1c92ac764d8bR171). The issue seems specific to the Drawer component and DesignKit page, this line could possibly be removed if we update the elements used in rendering these specific elements.
4. The `Theme` type is currently ambiguous, originally I had planned to create a more concrete type but that body of work was a larger effort than I had expected. I recommend deferring that to another ticket once this lands if possible.  
  









## Test Plan

1. Open the Determined application, using the Theme toggler which can be accessed by clicking your username in the top left corner. Toggle the theme between System, Light, and Dark mode. Ensure that the UI theme correctly changes between Light and Dark modes. 
2. Navigate to the `det/design` page. Toggle the theme between System, Light, and Dark mode by using the toggle in the top left corner. Ensure that the components and text on the page respond correctly to the Theme changes. 





## Commentary (optional)





## Checklist

  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.

## Ticket

WEB-1726
